algorithms/linfa-elasticnet/src/algorithm.rs

```diff
@@ -187,6 +306,88 @@ fn coordinate_descent<'a, F: Float>(
     (w, gap, n_steps)
 }
 
+fn block_coordinate_descent<'a, F: Float>(
+    x: ArrayView2<'a, F>,
+    y: ArrayView2<'a, F>,
+    tol: F,
+    max_steps: u32,
+    l1_ratio: F,
+    penalty: F,
+) -> (Array2<F>, F, u32) {
+    let n_samples = F::cast(x.shape()[0]);
+    let n_features = x.shape()[1];
+    let n_tasks = y.shape()[1];
+    // the parameters of the model
+    let mut w = Array2::<F>::zeros((n_features, n_tasks));
+    // the residuals: `Y - XW` (since W=0, this is just `Y` for now),
+    // the residuals are updated during the algorithm as the parameters change
+    let mut r = y.to_owned();
+    let mut n_steps = 0u32;
+    let norm_cols_x = x.map_axis(Axis(0), |col| col.dot(&col));
+    let mut gap = F::one() + tol;
+    let d_w_tol = tol;
+    let tol = tol * y.fold(F::zero(), |sum, &y_ij| sum + y_ij.powi(2));
```

Comment:
Instead of `fold` you can do `y.iter().map(|x| x*x).sum()`. Since `iter` is called before `map` it won't create a new array.  For 1D arrays it's even simpler since you can just call `y.dot(&y)` to dot product `y` with itself.

Reply:
You should also apply this change to all the other similar `fold` calls